runtime/lua/vim/lsp/codelens.lua

```diff
@@ -245,7 +244,10 @@ function M.refresh()
     return
   end
   active_refreshes[bufnr] = true
-  vim.lsp.buf_request(0, 'textDocument/codeLens', params)
+  vim.lsp.buf_request(0, 'textDocument/codeLens', params, function (...)
+    active_refreshes[bufnr] = nil
```

Comment:
This would clear it too early in the success case - it should wait until the lenses are all resolved. 

`M.on_codelens` needs to be called always even in the error case and https://github.com/neovim/neovim/blob/cc27540560672c1586e17a9e39512074e770b005/runtime/lua/vim/lsp/codelens.lua#L217 needs to be adjusted.

Could probably also use a test-case

Reply:
Ah, I figured it was only throttling the initial request.

> Could probably also use a test-case

Yeah I wanted to start with a failing test, but struggled a bit with how to pull it off as I couldn't find a way to mock/stub the RPC client interface. Any pointers?

Reply:
See https://github.com/neovim/neovim/blob/039d60bd9c4468bb53958943196390673569bd7b/test/functional/plugin/lsp_spec.lua#L2784-L2836

The test_name links to a fake-server implementation:

https://github.com/neovim/neovim/blob/039d60bd9c4468bb53958943196390673569bd7b/test/functional/plugin/lsp_spec.lua#L2792

https://github.com/neovim/neovim/blob/039d60bd9c4468bb53958943196390673569bd7b/test/functional/fixtures/fake-lsp-server.lua#L781-L793

You can create a new one that uses `expect_request` to send an error response to the client. 

Similar to here, (except that it should return the error instead of a regular result) 

https://github.com/neovim/neovim/blob/039d60bd9c4468bb53958943196390673569bd7b/test/functional/fixtures/fake-lsp-server.lua#L773-L775

Reply:
I pushed https://github.com/neovim/neovim/pull/18331/commits/fa35548343afe09c969574e58a50a31b8d14c57b. The test code got a bit messy as I was struggling a bit with the asynchronous nature of `vim.lsp.codelens.refresh()` (I'd rather not have to override and `vim.wait` for `vim.lsp.codelens.on_codelens`, if possible)